### PR TITLE
Add Organization Reconciler

### PR DESCRIPTION
**Pull request by glrf**
## Summary

This adds basic reconciliation of Organizations. This means it will mirror every change / creation / deletion of Organizations and OrganizationMembers.

Don't worry about the `2,491` added lines nearly half of it is `go.sum` and there is a lot of generated code / boilerplate. Interesting files are `controllers/organization_controller.go` and `keycloak/client.go`.

Out of scope for this PR and will follow in other PRs:

* **Group import**: Importing keycloak groups to organization
* **Improved Error Handling**: I'd like to handle some errors more gracefully and emit events. For example if adding a single user fails.
* **Deployment**
* **Resilience to Groupname change**: Currently if the name of a Keycloak group changes we assume the old one dissapeared. It would be nice to have some resilience by holding the id in the state and revert name changes. This however requires change in the control-api

## Checklist

- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [x] Update tests.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,3 +10,49 @@
 [build]: https://github.com/vshn/appuio-keycloak-adapter/actions?query=workflow%3ATest
 [releases]: https://github.com/vshn/appuio-keycloak-adapter/releases
 [codeclimate]: https://codeclimate.com/github/vshn/appuio-keycloak-adapter
+
+The [APUUiO Control API](https://github.com/appuio/control-api) enables self-service for  [APPUiO Cloud](https://appuio.cloud).
```
Comment (simu) [inline README.md:14]:
```suggestion
The [APPUiO Control API](https://github.com/appuio/control-api) enables self-service for  [APPUiO Cloud](https://appuio.cloud).
```

File: keycloak/client.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+		Search: &name,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	var group *gocloak.Group
+	for i := range groups {
+		if *groups[i].Name == name {
+			group = groups[i]
```
Comment (simu) [inline keycloak/client.go:128]: Minor nitpick: should we break out of the loop here?

```suggestion
			group = groups[i]
			break;
```

Or directly return, like we do in `getUserByName()`

```suggestion
			return groups[i], nil
```

**Review by simu — APPROVED** [review]
LGTM overall, didn't look at the tests in detail

File: keycloak/client.go (hunk 2 above)
Comment (glrf) [reply keycloak/client.go:128]: Good point. That should work the same way as in `getUserByName` for consistency.

File: controllers/organization_controller.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+		userRefs = append(userRefs, controlv1.UserRef{
+			Name: u,
+		})
+	}
+	memb.Status.ResolvedUserRefs = userRefs
+	return r.Status().Update(ctx, memb)
+}
+
+func buildKeycloakGroup(org *orgv1.Organization, memb *controlv1.OrganizationMembers) keycloak.Group {
+	groupMem := []string{}
```
Comment (bastjan) [inline controllers/organization_controller.go:139]:
```suggestion
	groupMem := make([]string, 0, len(memb.Spec.UserRefs))
```

Not that it really matters...

File: controllers/organization_controller.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+		controllerutil.RemoveFinalizer(memb, orgFinalizer)
+		if err := r.Update(ctx, memb); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *OrganizationReconciler) updateOrganizationStatus(ctx context.Context, org *orgv1.Organization, memb *controlv1.OrganizationMembers, group keycloak.Group) error {
+	userRefs := []controlv1.UserRef{}
```
Comment (bastjan) [inline controllers/organization_controller.go:128]:
```suggestion
	userRefs := make([]controlv1.UserRef, 0, len(group.Members))
```

**Review by bastjan — APPROVED** [review]
I really like the clean organization controller tests 😊

**Pull request merged**